### PR TITLE
manage additional factors

### PR DESCRIPTION
**Pull request by Djazouli**
Signed-off-by: Gael Macherel <gael.macherel@artelys.com>

**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*

**What is the current behavior?** *(You can also link to an open issue here)*

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

**Review by geofjamg — COMMENTED** [review]
Looks good to me, the way it is implemented is very clear and simple.
I am just wondering if there is a better design for ContingencyContext. Marker interfaces are often not considered as a good practice, but not easy to do otherwise.

File: src/main/java/com/powsybl/openloadflow/sensi/SensitivityFactorReader.java [hunk 1]
```diff
@@ -13,11 +13,29 @@
  */
 public interface SensitivityFactorReader {
 
+    interface ContingencyContext { }
```
Comment (flo-dup) [inline src/main/java/com/powsybl/openloadflow/sensi/SensitivityFactorReader.java:16]: Coudn't we replace this marker interface by something like:
```
enum ContingencyContextType {
   ALL, NONE, SPECIFIC
}
class ContingencyContext {
   private final String contingencyId;
   private final ContingencyContextType contextType;
}
```
Not the prettiest design either I know, but at least this avoids casting and instance of.

File: src/main/java/com/powsybl/openloadflow/sensi/AbstractSensitivityAnalysis.java [hunk 2]
```diff
@@ -675,12 +676,45 @@ private static void checkRegulatingTerminal(Network network, String equipmentId)
...
+        private final List<LfSensitivityFactor> commonFactors = new ArrayList<>();
+
+        public List<LfSensitivityFactor> getAllFactors() {
+            List<LfSensitivityFactor> allFactors = new ArrayList<>(commonFactors);
+            allFactors.addAll(additionalFactorsNoContingency);
+            allFactors.addAll(additionalFactorsPerContingency.values().stream().flatMap(List::stream).collect(Collectors.toSet()));
+            return allFactors;
+        }
+
+        public List<LfSensitivityFactor> getFactorsForContingency(String contingencyId) {
```
Comment (flo-dup) [inline src/main/java/com/powsybl/openloadflow/sensi/AbstractSensitivityAnalysis.java:691]: Just a detail, but I think you could avoid passing null arguments to a function to have a specific behaviour, and instead call a specific method directly. This avoids a test in the method but also avoids to give a sense to the null value (here I admit that the sense is easily guessed though!).

**Review by flo-dup — COMMENTED** [review]
Neat implementation - as already said above! Just a couple of remarks

File: src/main/java/com/powsybl/openloadflow/sensi/SensitivityFactorReaderAdapter.java [hunk 3]
```diff
@@ -46,50 +46,54 @@ public SensitivityFactorReaderAdapter(Network network, SensitivityFactorsProvide
 
     @Override
     public void read(Handler handler) {
+        ContingencyContext commonContingencyContext = new ContingencyContext(ContingencyContextType.ALL, null);
```
Comment (flo-dup) [inline src/main/java/com/powsybl/openloadflow/sensi/SensitivityFactorReaderAdapter.java:49]: Just as small improvement, would be better to use 3 static methods to create ContingencyContexts (`ContingencyContext.createAllContingencyContext()`, `ContingencyContext.createNoneContingencyContext()`, `ContingencyContext.createSpecificContingencyContext(String id)`)

**Review by flo-dup — CHANGES_REQUESTED** [review]

**Review by geofjamg — APPROVED** [review]

**Pull request merged**